### ISSUE TITLE
cam_dev appears to be failing the Ensemble Consistency test

### ISSUE THREAD

**Issue by cacraigucar** [ESCOMP/CAM#883]
### What happened?

Email exchange which captures the information to date.

From @sjsprecious: As I mentioned in my PUMAS PR (https://github.com/ESCOMP/PUMAS/pull/56), a code update in cam6_3_116 seemed to fail the ensemble consistency test for a GPU run. I looked at the PR for cam6_3_116 (https://github.com/ESCOMP/CAM/pull/820) and there were some changes related to the PUMAS and cam_dev settings. I also saw some communications between you and Adam saying that something might not work for cam_dev at this moment? I am using the F2000dev compset and wonder whether you know any changes in this PR that may cause the failure here?

From @cacraigucar: To the best of our abilities, we expect that cam_dev should be working properly with this and all of the other CAM tags we've made recently.  The back and forth discussions were our making mods to make sure cam_dev was working properly because cam_dev is our workhorse setup for the release.  cam6_3_116 was a tag which was brought in to incorporate the namelist changes (among other things) that Cecile and the scientists had been making and validated as being what they wanted at the time.  It was designed to mimic a cam_dev run that Cecile had made which gave improved results.  This was an answer changing tag.

From @sjsprecious: Suggested by Jim, I performed another ensemble consistency test (ECT) by using cam6_3_116 and the intel/2023 compiler on Derecho's CPU, and the ECT failed as well compared to the baseline generated on Cheyenne with intel/[19.1.1.](http://19.1.0.1/) Thus something could be wrong in the "cam_dev" physics and Jim is helping dig into it further.

@adamrher @PeterHjortLauritzen @cecilehannay  - I want to make sure you are aware of this concern

### What are the steps to reproduce the bug?

Run the ensemble consistency test as described above

### What CAM tag were you using?

cam6_3_116 

### What machine were you running CAM on?

CISL machine (e.g. cheyenne)

### What compiler were you using?

Intel

### Will you be addressing this bug yourself?

No

**Comment by sjsprecious** [comment]
Suggested by @briandobbins , I performed another test simulation on Derecho's CPU with the intel/2023 compiler and 36 MPI ranks per node (i.e., the same PE layout as the one on Cheyenne). The new simulation still failed the ECT.

**Comment by sjsprecious** [comment]
@jedwards4b @cacraigucar I just found out the issue about the intel compiler:
- The `cam6_3_116` tag uses the `ccs_config_cesm0.0.72` tag (https://github.com/ESCOMP/CAM/blob/cam6_3_116/Externals.cfg#L2) for Derecho's machine configuration, where the `-no-fma` option is somehow missing for the intel compiler (https://github.com/ESMCI/ccs_config_cesm/blob/ccs_config_cesm0.0.72/machines/cmake_macros/intel_derecho.cmake#L3). This is inconsistent with the settings on Cheyenne, where we always turn off FMA.

After adding the `-no-fma` option back for Derecho, I could confirm that the ensemble consistency test now passes for the `cam6_3_116` tag with the intel/2023.0 compiler on Derecho. Still no clues about why NVHPC fails the ECT.

I can issue a quick PR on the `ccs_config_cesm` repo to update the settings for the intel compiler on Derecho if it makes sense to you.

**Comment by jedwards4b** [comment]
@sjsprecious Good catch - thank you.

**Comment by sjsprecious** [comment]
More updates: After I switched `use_gw_front` and `use_gw_convect_dp` back to `False` (the values used in `cam6_3_115`), and re-generated the baseline on Cheyenne with the intel compiler, the ECT now passed for the GPU run on Casper with the nvhpc compiler.

I am continuing to check whether I have to reset both variables or only one of them matters. At the same time, it will be great if someone could comment what these two variables are used for.

**Comment by adamrher** [comment]
sorry to be late on picking up this issue, but in #820, all the cam_dev baselines should've failed due to adding a switch to turn on ```use_gw_front``` and ```use_gw_convect_dp``` with cam_dev in bld/build-namelist. Before this PR those namelists were false for cam_dev, I believe. So if this ensemble consistency test wasn't run in that PR, and this is the first time it's been run since, it should fail wrt to the old baseline.

**Comment by sjsprecious** [comment]
Thanks @adamrher for your information. Just to clarify: I generated a new baseline for `cam6_3_116` when I performed the ensemble consistency test, not comparing to the old baseline.

**Comment by adamrher** [comment]
Oh ok. Failing with new baselines, gotchya. To answer your original question, those two namelists turn on the convective and frontal gravity options in the gravity wave scheme ```gw_tend```.

**Comment by sjsprecious** [comment]
Thanks Adam for your information. Will look at the `gw_tend` scheme later.

**Comment by sjsprecious** [comment]
Ok, I think I am getting closer:

- By only resetting `use_gw_convect_dp` back to `False` (the values used in `cam6_3_115`), and re-generated the baseline on Cheyenne with the intel compiler, the ECT also passed for the GPU run on Casper with the nvhpc/22.2 compiler.

Therefore, I think either `use_gw_convect_dp` turns on some code that is problematic or there is another compiler bug related to NVHPC. I tried the latest `nvhpc/23.7` compiler on Derecho and the ECT still failed for the default `cam6_3_116`  tag. Thus if it is indeed a compiler bug, it is not fixed yet in the latest NVHPC release.

I am looking at the code section turned on by `use_gw_convect_dp` (https://github.com/ESCOMP/CAM/blob/cam_development/src/physics/cam/gw_drag.F90#L1457) but I have not found anything suspicious so far. I will continue to check the sibling subroutines and it is much appreciated if someone could also provide additional thoughts here.

**Comment by adamrher** [comment]
@JulioTBacmeister making you aware that there **may** be an issue with ```use_gw_convect_dp```, and so if you have anything to add here for Jian, he'd appreciate it. Although it's not clear if this is a bug in the code or a problem with the nvhpc compiler (see above).

**Comment by sjsprecious** [comment]
@cacraigucar @adamrher @jedwards4b Ok, I think I find out the culprit (at least one of them):

- intel and nvhpc compilers behave diffrerently for the `eoshift` function.

By using the example modified from https://gcc.gnu.org/onlinedocs/gfortran/EOSHIFT.html:

```
program test_eoshift
    integer, dimension(3,3) :: a
    a = reshape( (/ 1, 2, 3, 4, 5, 6, 7, 8, 9 /), (/ 3, 3 /))
    print '(3i3)', a(1,:)
    print '(3i3)', a(2,:)
    print '(3i3)', a(3,:)
    a = EOSHIFT(a, SHIFT=0, BOUNDARY=-5, DIM=2)
    print *
    print '(3i3)', a(1,:)
    print '(3i3)', a(2,:)
    print '(3i3)', a(3,:)
end program test_eoshift
``` 

Intel compiler returns:
```
  1  4  7
  2  5  8
  3  6  9
```
which I think is correct since `shift = 0`.

However, nvhpc compiler returns:
```
 -5 -5 -5
 -5 -5 -5
 -5 -5 -5
```
which is incorrect compared to the result above.

We may have two solutions for this issue:

1. Send this example to NVIDIA and ask them to fix the compiler bug in their future release.
2. Replace `eoshift` function by a different implementation that both intel and nvhpc compilers return the correct answers.

Let me know how you want me to proceed.

**Comment by jedwards4b** [comment]
Can we do both?   Please put this in an email to cislhelp so that they can file a bug report on the nvhpc compiler and see if we can find a suitable workaround for the nvhpc compiler - this can either be a replacement for EOSHIFT for all compilers or a #ifdef CPRPGI clause.

**Comment by cacraigucar** [comment]
Whatever we do, we need to make sure it covers all of the places that EOSHIFT is used.  CLUBB (both the main version and the older version used in SPCAM) have several places where it is used.

**Comment by sjsprecious** [comment]
Thanks @jedwards4b . I just sent an email to NVIDIA's staff about this issue. Hopefully they can fix it soon.

**Comment by adamrher** [comment]
Since there is only one call to ```eoshift``` in all of the cam physics, we could just replace it with something like:

```
  tmptau0 = tau0
  do m=1,n
    if ((m+shift) > 0) then
      tau0(m+shift) = tmptau0(m)
    else
      tau0(n+(m+shift)) = tmptau0(m)
    end if
  end do
```

@JulioTBacmeister would that work? Not so sure about the statement in the else block.

**Comment by sjsprecious** [comment]
Thanks @cacraigucar . This is very helpful information. Maybe a fix from the compiler's end is a better solution in case we miss an `EOSHIFT` function call somewhere.

**Comment by sjsprecious** [comment]
> Since there is only one call to `eoshift` in all of the cam physics, we could just replace it with something like:
> 
> ```
>   tmptau0 = tau0
>   do m=1,n
>     if ((m+shift) > 0) then
>       tau0(m+shift) = tmptau0(m)
>     else
>       tau0(n+(m+shift)) = tmptau0(m)
>     end if
>   end do
> ```
> 
> @JulioTBacmeister would that work? Not so sure about the statement in the else block.
> 
> [edit -- per Cheryl, this is not the only place eoshift is used. And this implementation won't work generally; just in the 1D case]

Thanks @adamrher for providing an alternative. You are right that this implementation only works for 1-D case and it may not be efficient compared to the `EOSHIFT` function. I have sent an email to NVIDIA about this bug and hopefully they could provide a fix on the compiler's end soon.

**Comment by JulioTBacmeister** [comment]
I am going through Adam's code snippet, but I think EOSHIFT should really be fixed at the compiler level. It seems Cheryl says it is used in other places besides gw_convect

**Comment by cponder** [comment]
Is this EOSHIFT problem with the 23.7 compiler?

**Comment by sjsprecious** [comment]
@cponder : yes, I tested `nvhpc/23.7` and it produced the wrong result above.

**Comment by sjsprecious** [comment]
update: @cponder has helped file a bug report to NVIDIA.

**Comment by sjsprecious** [comment]
According to @cponder , the NVHPC compiler bug related to `EOSHIFT` is expected to be fixed in the release of `nvhpc/24.1`, which will be available in February 2024.

**Comment by sjsprecious** [comment]
@jedwards4b @cacraigucar @adamrher An update about this issue:
- The `EOSHIFT` problem with the NVHPC compiler is fixed in the latest `nvhpc/24.1` release.
- There is a new compilation error when using `nvhpc/23.11` or newer version to build CAM, and the error comes from the CICE build. More details could be found at https://github.com/EarthWorksOrg/EarthWorks/issues/23. There is a potential workaround provided in that GitHub link, too.

**Comment by sjsprecious** [comment]
Closed by https://github.com/ESMCI/ccs_config_cesm/pull/163.
